src/common/utilities/FormulaEvaluation.ts

```diff
@@ -0,0 +1,641 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { sp } from "@pnp/sp";
+import { IContext } from "../Interfaces";
+import { ASTNode, ArrayLiteralNode, Token, TokenType, ValidFuncNames } from "./FormulaEvaluation.types";
+
+export class FormulaEvaluation {
+    private webUrl: string;
+    private _meEmail: string;
+    
+    constructor(context?: IContext, webUrlOverride?: string) {
+        if (context) {
+            sp.setup({ pageContext: context.pageContext });
+            this._meEmail = context.pageContext.user.email;
+        }
+        this.webUrl = webUrlOverride || context?.pageContext.web.absoluteUrl || '';
+    }
+
+    /** Evaluates a formula expression and returns the result, with optional context object for variables */
+    public evaluate(expression: string, context: { [key: string]: any } = {}): any {
+        context['me'] = this._meEmail;
+        context['today'] = new Date();
+        const tokens: Token[] = this.tokenize(expression, context);
+        const postfix: Token[] = this.shuntingYard(tokens);
+        const ast: ASTNode = this.buildAST(postfix);
+        return this.evaluateASTNode(ast, context);
+    }
+
+    /** Tokenizes an expression into a list of tokens (primatives, operators, variables, function names, arrays etc) */
+    public tokenize(expression: string, context: { [key: string]: any } = {}): Token[] {
+        // Each pattern captures a different token type
+        // and are matched in order
+        const patterns: [RegExp, TokenType][] = [
```

Comment:
Make a const outside of the class?

Reply:
`patterns` and `operatorTypes` (the array featuring all of the various operators) are now `const` outside of the class 👍